h2o-core/src/main/java/water/webserver/H2OHttpViewImpl.java

```diff
@@ -0,0 +1,194 @@
+package water.webserver;
+
+import org.apache.commons.io.IOUtils;
+import water.ExtensionManager;
+import water.api.DatasetServlet;
+import water.api.NpsBinServlet;
+import water.api.PostFileServlet;
+import water.api.PutKeyServlet;
+import water.api.RequestServer;
+import water.server.ServletUtils;
+import water.util.Log;
+import water.webserver.iface.H2OHttpConfig;
+import water.webserver.iface.H2OHttpView;
+import water.webserver.iface.RequestAuthExtension;
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+
+/**
+ * This is intended to be a singleton per H2O node.
+ */
+public class H2OHttpViewImpl implements H2OHttpView {
+  private static volatile boolean _acceptRequests = false;
+  private final H2OHttpConfig config;
+  private static final LinkedHashMap<String, Class<? extends HttpServlet>> SERVLETS = new LinkedHashMap<>();
+  static {
+    SERVLETS.put("/3/NodePersistentStorage.bin/*", NpsBinServlet.class);
+    SERVLETS.put("/3/PostFile.bin", PostFileServlet.class);
+    SERVLETS.put("/3/PostFile", PostFileServlet.class);
+    SERVLETS.put("/3/DownloadDataset", DatasetServlet.class);
+    SERVLETS.put("/3/DownloadDataset.bin", DatasetServlet.class);
+    SERVLETS.put("/3/PutKey.bin", PutKeyServlet.class);
+    SERVLETS.put("/3/PutKey", PutKeyServlet.class);
+    SERVLETS.put("/", RequestServer.class);
+  }
+
+  public H2OHttpViewImpl(H2OHttpConfig config) {
+    this.config = config;
+  }
+
+  public void acceptRequests() {
+    _acceptRequests = true;
+  }
+
+  /**
+   * @return URI scheme
+   */
+  public String getScheme() {
+    return config.jks == null ? "http" : "https";
+  }
+
+  @Override
+  public LinkedHashMap<String, Class<? extends HttpServlet>> getServlets() {
+    return SERVLETS;
+  }
+
+  @Override
+  public boolean authenticationHandler(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    if (!config.loginType.isJaas()) {
+      //TODO question: for LoginType.HASH, does this equal not adding this handler at all? if so, it might be better doing it that way
+      return false;
+    }
+
+    final String loginName = request.getUserPrincipal().getName();
+    if (loginName.equals(config.user_name)) {
+      return false;
+    }
+    Log.warn("Login name (" + loginName + ") does not match cluster owner name (" + config.user_name + ")");
+    ServletUtils.sendResponseError(response, HttpServletResponse.SC_UNAUTHORIZED, "Login name does not match cluster owner name");
+    return true;
+  }
+
+  @Override
+  public void gateHandler(HttpServletRequest request, HttpServletResponse response) {
+    ServletUtils.startRequestLifecycle();
+    while (!_acceptRequests) {
```

Comment:
I see a simple way of avoiding the loop via:

```java
      if(!_acceptRequests) {
        try {
          wait();
          _acceptRequests = true;
        } catch (InterruptedException e) {
          H2O.fail(e.getMessage(), e);
        }
      }
```

Only the `acceptRequests()` method needs modification (see above, line 50).

Reply:
H2O.fail would shut down H2O, is that intended. I would keep the simplicity of the original solution.